doc/ltsmc.1
.\" Copyright (c) 2017, Jörg Behrendt
.\"
.\" %%%LICENSE_START(GPLv2+_DOC_FULL)
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, see
.\" <http://www.gnu.org/licenses/>.
.\" %%%LICENSE_END

.TH "ltsmc" "1" "1. June 2017" "Thomas Stibor <t.stibor@gsi.de>, Jörg Behrendt <j.behrendt@gsi.de>" "version 0.5.0"

.SH NAME
ltsmc is a console client for archiving, retrieving, deleting and querying files and directories on a TSM server

.SH SYNOPSIS
ltsmc ACTION OPTION... PATH

.SH DESCRIPTION
.SS
.BR ACTIONS:
.TP
.BR \-\-archive
Archive files and directories on TSM server.
.TP
.BR \-\-query
Query TSM server for files and directories and print the result.
.TP
.BR \-\-retrieve
Query TSM server for files and directories, and retrieve matched files and directories for storing them locally.
.TP
.BR \-\-delete
Query TSM server for files and directories, and delete matched files and directories on the TSM server.
.TP
.BR \-\-pipe
Archive data coming from a pipe, e.g. echo "archive me" | ltsmc --pipe <OPTIONS> /archive.me
.TP
.BR \-c ", " \-\-checksum
Calculate crc32 checksum of local files, for verifying that e.g. local and archives files
have equal crc32 checksums.
.SS
.BR OPTIONS:
Mandatory arguments to long options are mandatory for short options too except for verbose option.
.PP
REQUIRED
.TP
.BR \-n ", " \-\-node =\fINAME\fR
The name of the TSM node which is registered on the TSM server.
.TP
.BR \-p ", " \-\-password =\fIPASSWORD\fR
The password of the TSM node which is registered on the TSM server.
.TP
.BR \-s ", " \-\-servername =\fINAME\fR
The TSM servername to connect to. Name will be resolved to an actual ip address within the dsm.sys file.
.PP
OPTIONAL
.TP
.BR \-x ", " \-\-prefix
The directory prefix is the directory where all retrieved files and sub-directories will be saved to, i.e. the top of the retrieval tree. For example, --prefix '/tmp' assures that archived data stored at fs=/lustre/nyx,hl=/cs/turing,ll=/pub.tex is retrieved as file /tmp/lustre/nyx/cs/turing/pub.tex.
.TP
.BR \-r ", " \-\-recursive
Archive directory recursively by also processing each sub-directory.
.TP
.BR \-t ", " \-\-sort =\fIascending\fR|\fIdescending\fR|\fIrestore\fR
Query action will list archived objects sorted by date in \fIascending\fR or \fIdescending\fR order, or in optimal \fIrestore\fR order.
.TP
.BR \-l ", " \-\-latest
Query, restore and delete action will only select the latest version of the stored objects.
.TP
.BR \-f ", " \-\-fsname =\fIPATH\fR
Name of a file system, disk drive, or any other high-level qualifier that groups related data together, e.g. '/lustre'. The TSM filespace name is used for archiving, retrieving, querying and deleting and has default value '/'. Note the filespace name must start with character '/'.
.TP
.BR \-d ", " \-\-description =\fISTRING\fR
If in archive mode the description will be stored in association with the TSM objects onto the TSM server. By default an empty string is used.
Otherwise the description will be used to select only the query results matching the description before continuing with the actual action and can contain the wildcards '*' or '?'. By default any description '*' will be selected.
.TP
.BR \-o ", " \-\-owner =\fINAME\fR
If in archive mode the owner's name  will be stored with the file objects onto the TSM server. By default an empty string is used.
Otherwise the name will be used to filter the queries by owner and the wildcards '*' or '?' are allowed. By default all owners will be selected.
.TP
.BR \-v ", " \-\-verbose =\fIerror\fR|\fIwarn\fR|\fImessage\fR|\fIinfo\fR|\fIdebug\fR
Causes ltsmc to be more verbose in printing messages. Default is \fImessage\fR.
.TP
.BR \-h ", " \-\-help
Display help and exit.
.SS
.BR PATH:
.PP
If not in archive mode a TSM object description is expected. The description is split into the low-level (ll) name which reflects the actual file name and a high-level (hl) name which reflects the path.
This split will occur at the last occurency of the path seperator '/' in the string. In addition wildcard characters '*' and '?' can be used to match multiple TSM objects, e.g.:
.TP
PATH='*'
hl='/' and ll='*' \- matches any file in path '/'
.TP
PATH='*/*'
hl='*' and ll='/*' \- matches any file in any path
.TP
PATH='/home/user/files/test.txt'
hl='/home/user/files' and ll='/test.txt' \- matches single file
.TP
PATH='/home/user/*.txt'
hl="/home/user" and ll='/*.txt' \- matches any file ending with '.txt' in path '/home/user'
.TP
PATH='/home/*/files/??.txt'
hl='/home/*/files' and ll='/??.txt' \- matches all files with two characters and ending with '.txt' in all paths matching '/home/*/files'.
.PP
When in archive mode a path to an existing file or directory is required which should be stored on the tsm server. If path points to an directory all files in it will be archived. See additional options to store directories recursivly.
Note, the PATH parameter must be in quotation marks, when wildcard character '*' is provided, otherwise the expression is shell expanded.

.SH FILES
File 'dsm.sys' contains information to look up the tsm server. By default it is located at '/opt/tivoli/tsm/client/api/bin64/dsm.sys'.
It should contain at least the required options SERVERNAME, NODENAME and TCPSERVERADDRESS.

.SH BUGS
Please report bugs to <http://github.com/tstibor/ltsm/issues>

.SH SEE ALSO
Github project at <http://github.com/tstibor/ltsm>


src/ltsmc.c
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 only,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License version 2 for more details (a copy is included
 * in the LICENSE file that accompanied this code).
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * Copyright (c) 2016, 2017, Thomas Stibor <t.stibor@gsi.de>
 * 			     Jörg Behrendt <j.behrendt@gsi.de>
 */

#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <getopt.h>
#include <sys/stat.h>
#include <zlib.h>
#include "tsmapi.h"

struct options {
	int o_archive;
	int o_retrieve;
	int o_query;
	int o_delete;
	int o_pipe;
	int o_verbose;
	int o_latest;
	int o_recursive;
	int o_checksum;
	int o_sort;
	char o_servername[DSM_MAX_SERVERNAME_LENGTH + 1];
	char o_node[DSM_MAX_NODE_LENGTH + 1];
	char o_owner[DSM_MAX_OWNER_LENGTH + 1];
	char o_password[DSM_MAX_VERIFIER_LENGTH + 1];
	char o_fsname[DSM_MAX_FSNAME_LENGTH + 1];
	char o_fstype[DSM_MAX_FSTYPE_LENGTH + 1];
	char o_desc[DSM_MAX_DESCR_LENGTH + 1];
};

struct options opt = {
	.o_verbose = API_MSG_NORMAL,
	.o_servername = {0},
	.o_node = {0},
	.o_owner = {0},
	.o_password = {0},
	.o_fsname = {0},
	.o_fstype = {0},
	.o_desc = {0},
	.o_checksum = 0,
	.o_sort = SORT_NONE
};

static void usage(const char *cmd_name, const int rc)
{
	dsmApiVersionEx libapi_ver = get_libapi_ver();
	dsmAppVersion appapi_ver = get_appapi_ver();

	fprintf(stdout, "usage: %s [options] <files|directories|wildcards>\n"
		"\t--archive\n"
		"\t--retrieve\n"
		"\t--query\n"
		"\t--delete\n"
		"\t--pipe\n"
		"\t-l, --latest [retrieve object with latest timestamp when multiple exists]\n"
		"\t-x, --prefix [retrieve prefix directory]\n"
		"\t-r, --recursive [archive directory and all sub-directories]\n"
		"\t-t, --sort={ascending, descending, restore} [sort query in date or restore order]\n"
		"\t-f, --fsname <string> [default: '/']\n"
		"\t-d, --description <string>\n"
		"\t-n, --node <string>\n"
		"\t-o, --owner <string>\n"
		"\t-p, --password <string>\n"
		"\t-s, --servername <string>\n"
		"\t-v, --verbose {error, warn, message, info, debug} [default: message]\n"
		"\t-c, --checksum <file>\n"
		"\t-h, --help\n"
		"\nIBM API library version: %d.%d.%d.%d, "
		"IBM API application client version: %d.%d.%d.%d\n"
		"version: %s © 2017 by Thomas Stibor <t.stibor@gsi.de>,"
		" Jörg Behrendt <j.behrendt@gsi.de>\n",
		cmd_name,
		libapi_ver.version, libapi_ver.release, libapi_ver.level,
		libapi_ver.subLevel,
		appapi_ver.applicationVersion, appapi_ver.applicationRelease,
		appapi_ver.applicationLevel, appapi_ver.applicationSubLevel,
		PACKAGE_VERSION);
	exit(rc);
}

static void sanity_arg_check(const char *argv)
{
	uint8_t count = 0;
	count = opt.o_archive  == 1 ? count + 1 : count;
	count = opt.o_retrieve == 1 ? count + 1 : count;
	count = opt.o_delete   == 1 ? count + 1 : count;
	count = opt.o_query    == 1 ? count + 1 : count;
	count = opt.o_checksum == 1 ? count + 1 : count;
	count = opt.o_pipe     == 1 ? count + 1 : count;

	if (count == 0) {
		fprintf(stdout, "missing argument --archive, --retrieve,"
			" --query, --delete, --pipe or --checksum\n\n");
		usage(argv, 1);
	} else if (count != 1) {
		printf("multiple incompatible arguments"
		       " --archive, --retrieve, --query, --delete, --pipe or "
		       "--checksum\n\n");
		usage(argv, 1);
	}

	/* There are no additional required arguments when
	   --checksum is chosen. */
	if (opt.o_checksum)
		return;

	/* Required arguments. */
	if (!strlen(opt.o_node)) {
		fprintf(stdout, "missing argument -n, --node <string>\n\n");
		usage(argv, 1);
	} else if (!strlen(opt.o_password)) {
		fprintf(stdout, "missing argument -p, --password <string>\n\n");
		usage(argv, 1);
	} else if (!strlen(opt.o_servername)) {
		fprintf(stdout, "missing argument -s, --servername "
			"<string>\n\n");
		usage(argv, 1);
	} else if (!strlen(opt.o_fsname)) {
		strncpy(opt.o_fsname, DEFAULT_FSNAME, strlen(DEFAULT_FSNAME));
	}
}

static int parseopts(int argc, char *argv[])
{
	struct option long_opts[] = {
		{"archive",	      no_argument, &opt.o_archive,     1},
		{"retrieve",	      no_argument, &opt.o_retrieve,    1},
		{"query",	      no_argument, &opt.o_query,       1},
		{"delete",	      no_argument, &opt.o_delete,      1},
		{"pipe",              no_argument, &opt.o_pipe,        1},
		{"latest",	      no_argument, 0,		     'l'},
		{"recursive",	      no_argument, 0,		     'r'},
		{"sort",	required_argument, 0,		     't'},
		{"fsname",	required_argument, 0,		     'f'},
		{"description", required_argument, 0,		     'd'},
		{"node",	required_argument, 0,		     'n'},
		{"owner",	required_argument, 0,		     'o'},
		{"password",	required_argument, 0,		     'p'},
		{"servername",	required_argument, 0,		     's'},
		{"verbose",	required_argument, 0,		     'v'},
		{"prefix",	required_argument, 0,		     'x'},
		{"checksum",	      no_argument, 0,		     'c'},
		{"help",	      no_argument, 0,		     'h'},
		{0, 0, 0, 0}
	};

	int c;
	while ((c = getopt_long(argc, argv, "lrt:f:d:n:o:p:s:v:x:ch",
				long_opts, NULL)) != -1) {
		switch (c) {
		case 'l': {
			opt.o_latest = 1;
			break;
		}
		case 'r': {
			opt.o_recursive = 1;
			set_recursive(bTrue);
			break;
		}
		case 't': {
			if (OPTNCMP("none", optarg))
				opt.o_sort = SORT_NONE;
			else if (OPTNCMP("ascending", optarg))
				opt.o_sort = SORT_DATE_ASCENDING;
			else if (OPTNCMP("descending", optarg))
				opt.o_sort = SORT_DATE_DESCENDING;
			else if (OPTNCMP("restore", optarg))
				opt.o_sort = SORT_RESTORE_ORDER;
			else {
				fprintf(stdout, "wrong argument for -t, "
					"--sort '%s'\n", optarg);
				usage(argv[0], 1);
			}
			break;
		}
		case 'f': {
			strncpy(opt.o_fsname, optarg, DSM_MAX_FSNAME_LENGTH);
			break;
		}
		case 'd': {
			strncpy(opt.o_desc, optarg, DSM_MAX_DESCR_LENGTH);
			break;
		}
		case 'n': {
			strncpy(opt.o_node, optarg, DSM_MAX_NODE_LENGTH);
			break;
		}
		case 'o': {
			strncpy(opt.o_owner, optarg, DSM_MAX_OWNER_LENGTH);
			break;
		}
		case 'p': {
			strncpy(opt.o_password, optarg,
				DSM_MAX_VERIFIER_LENGTH);
			break;
		}
		case 's': {
			strncpy(opt.o_servername, optarg,
				DSM_MAX_SERVERNAME_LENGTH);
			break;
		}
		case 'v': {
			if (OPTNCMP("error", optarg))
				opt.o_verbose = API_MSG_ERROR;
			else if (OPTNCMP("warn", optarg))
				opt.o_verbose = API_MSG_WARN;
			else if (OPTNCMP("message", optarg))
				opt.o_verbose = API_MSG_NORMAL;
			else if (OPTNCMP("info", optarg))
				opt.o_verbose = API_MSG_INFO;
			else if (OPTNCMP("debug", optarg))
				opt.o_verbose = API_MSG_DEBUG;
			else {
				fprintf(stdout, "wrong argument for -v, "
					"--verbose '%s'\n", optarg);
				usage(argv[0], 1);
			}
			api_msg_set_level(opt.o_verbose);
			break;
		}
		case 'x': {
			set_prefix(optarg);
			break;
		}
		case 'c': {
			opt.o_checksum = 1;
			break;
		}
		case 'h': {
			usage(argv[0], 0);
			break;
		}
		case 0: {
			break;
		}
		default:
			return -EINVAL;
		}
	}

	sanity_arg_check(argv[0]);

	return 0;
}

static int calc_crc32sum(const char *filename, uint32_t *crc32result)
{
	int rc;
	FILE *file;
	size_t cur_read;
	uint32_t crc32sum = 0;
	unsigned char buf[TSM_BUF_LENGTH] = {0};

	file = fopen(filename, "r");
	if (file == NULL) {
		rc = errno;
		CT_ERROR(rc, "fopen failed on '%s'", filename);
		return rc;
	}

	do {
		cur_read = fread(buf, 1, TSM_BUF_LENGTH, file);
		if (ferror(file)) {
			CT_ERROR(EIO, "fread failed on '%s'", filename);
			break;
		}
		crc32sum = crc32(crc32sum, (const unsigned char *)buf,
				 cur_read);

	} while (!feof(file));


	rc = fclose(file);
	if (rc) {
		rc = errno;
		CT_ERROR(rc, "fclose failed on '%s'", filename);
		return rc;
	}

	*crc32result = crc32sum;
	return rc;
}

int main(int argc, char *argv[])
{
	int rc;
	api_msg_set_level(opt.o_verbose);
	rc = parseopts(argc, argv);
	if (rc) {
		CT_WARN("try '%s --help' for more information", argv[0]);
		return 1;
	}

	size_t num_files_dirs = 0;
	char **files_dirs_arg = NULL;

	/* file or directory parameter */
	if (optind < argc) {
		size_t i = 0;
		num_files_dirs = argc - optind;
		files_dirs_arg = malloc(sizeof(char *) * num_files_dirs);
		while (optind < argc) {
			size_t len = strlen(argv[optind]);
			files_dirs_arg[i] = calloc(len + 1, sizeof(char));
			(files_dirs_arg[i])[len] = '\0';
			strncpy(files_dirs_arg[i++], argv[optind++], len);
		}
	}

	if (opt.o_checksum) {

		if (num_files_dirs == 0) {
			fprintf(stdout, "missing argument <files>\n");
			usage(argv[0], 1);
		}

		uint32_t crc32sum = 0;
		for (size_t i = 0; i < num_files_dirs &&
			     files_dirs_arg[i]; i++) {
			rc = calc_crc32sum(files_dirs_arg[i], &crc32sum);
			if (rc)
				CT_WARN("calculation of crc32 for '%s' failed",
					files_dirs_arg[i]);
			else
				fprintf(stdout, "crc32: "
					"0x%08x (%010u), file: '%s'\n", crc32sum,
					crc32sum, files_dirs_arg[i]);
		}
		goto cleanup;
	}

	struct login_t login;
	login_fill(&login, opt.o_servername,
		   opt.o_node, opt.o_password,
		   opt.o_owner, LINUX_PLATFORM,
		   opt.o_fsname, DEFAULT_FSTYPE);

	struct session_t session;
	bzero(&session, sizeof(session));

	session.qtable.multiple = opt.o_latest == 1 ? bFalse : bTrue;
	session.qtable.sort_by = opt.o_sort;

	rc = tsm_init(DSM_SINGLETHREAD);
	if (rc)
		goto cleanup_tsm;

	if (opt.o_pipe) {
		if (num_files_dirs == 0) {
			fprintf(stdout, "missing argument <files>\n");
			usage(argv[0], 1);
		} else if (num_files_dirs > 1) {
			fprintf(stdout, "too many arguments <files>\n");
			usage(argv[0], 1);
		}

		rc = tsm_fopen(opt.o_fsname, files_dirs_arg[0], opt.o_desc,
			       &login, &session);
		if (rc) {
			tsm_cleanup(DSM_SINGLETHREAD);
			goto cleanup;
		}

		char buf[TSM_BUF_LENGTH] = {0};
		size_t size;
		do {
			size = fread(buf, 1, TSM_BUF_LENGTH, stdin);
			if (ferror(stdin)) {
				session.tsm_file->err = EIO;
				CT_ERROR(EIO, "fread failed");
				break;
			}
			if (size == 0)
				break;
			rc = tsm_fwrite(buf, 1, size, &session);
			if (rc < 0) {
				CT_ERROR(errno, "tsm_fwrite failed");
				break;
			}
		} while (!feof(stdin));

		rc = tsm_fclose(&session);
		if (rc)
			CT_ERROR(errno, "tsm_fclose failed");

		tsm_cleanup(DSM_SINGLETHREAD);
		goto cleanup;
	}

	rc = tsm_connect(&login, &session);
	if (rc)
		goto cleanup_tsm;

	rc = tsm_query_session(&session);
	if (rc)
		goto cleanup_tsm;

	/* Handle operations on files and directories resp. */
	for (size_t i = 0; i < num_files_dirs &&
		     files_dirs_arg[i]; i++) {
		if (opt.o_query)
			rc = tsm_query_fpath(opt.o_fsname, files_dirs_arg[i],
					     opt.o_desc, &session);
		else if (opt.o_retrieve)
			rc = tsm_retrieve_fpath(opt.o_fsname, files_dirs_arg[i],
						opt.o_desc, -1, &session);
		else if (opt.o_delete)
			rc = tsm_delete_fpath(opt.o_fsname, files_dirs_arg[i],
					      &session);
		else if (opt.o_archive)
			rc = tsm_archive_fpath(opt.o_fsname,
					       files_dirs_arg[i],
					       opt.o_desc, -1, NULL, &session);
		if (rc)
			goto cleanup_tsm;
	}

cleanup_tsm:
	tsm_disconnect(&session);
	tsm_cleanup(DSM_SINGLETHREAD);

cleanup:
	if (files_dirs_arg) {
		for (size_t i = 0; i < num_files_dirs; i++) {
			if (files_dirs_arg[i])
				free(files_dirs_arg[i]);
		}
		free(files_dirs_arg);
	}

	return rc;
}
